plonky2/src/plonk/config.rs

```diff
@@ -46,35 +54,50 @@ pub trait Hasher<F: RichField>: Sized + Copy + Debug + Eq + PartialEq {
 
     /// Hash a message without any padding step. Note that this can enable length-extension attacks.
     /// However, it is still collision-resistant in cases where the input has a fixed length.
-    fn hash_no_pad(input: &[F]) -> Self::Hash;
+    fn hash_no_pad(input: &[F]) -> Self::Hash {
+        Self::hash_no_pad_iter(input.iter().copied())
+    }
+
+    /// Hash a message without any padding step. Note that this can enable length-extension attacks.
+    /// However, it is still collision-resistant in cases where the input has a fixed length.
+    fn hash_no_pad_iter<I: IntoIterator<Item = F>>(input: I) -> Self::Hash;
 
     /// Pad the message using the `pad10*1` rule, then hash it.
```

Comment:
Btw, can we pad with `pad10*` rule instead?  That one is one element shorter on average.

(I don't think this is something introduced in this PR.  So we can solve it in a new PR, too.)

Reply:
As far as I know we totally can. We could even do a rule `1(0*1)?`, padding with `1`, `11`, or `10*1` as required.

This would break compatibility with any existing code because it changes the hash, so we probably can't upstream it.